src/io/kqueue.rs

```diff
@@ -0,0 +1,558 @@
+
+        /// TODO
+        const EOF = c::EV_EOF as _;
+
+        /// TODO
+        const ERROR = c::EV_ERROR as _;
+    }
+}
+
+bitflags::bitflags! {
+    /// The flags for a virtual node event.
+    pub struct VnodeEvents : u32 {
+        /// The file was deleted.
+        const DELETE = c::NOTE_DELETE;
+
+        /// The file was written to.
+        const WRITE = c::NOTE_WRITE;
+
+        /// The file was extended.
+        const EXTEND = c::NOTE_EXTEND;
+
+        /// The file had its attributes changed.
+        const ATTRIBUTES = c::NOTE_ATTRIB;
+
+        /// The file was renamed.
+        const RENAME = c::NOTE_RENAME;
+
+        /// Access to the file was revoked.
+        const REVOKE = c::NOTE_REVOKE;
+    }
+}
+
+#[cfg(feature = "process")]
+bitflags::bitflags! {
+    /// The flags for a process event.
+    pub struct ProcessEvents : u32 {
+        /// The process exited.
+        const EXIT = c::NOTE_EXIT;
+
+        /// The process forked itself.
+        const FORK = c::NOTE_FORK;
+
+        /// The process executed a new process.
+        const EXEC = c::NOTE_EXEC;
+
+        /// TODO
+        const TRACK = c::NOTE_TRACK;
+    }
+}
+
+#[cfg(any(
+    target_os = "macos",
+    target_os = "ios",
+    target_os = "watchos",
+    target_os = "dragonfly",
+    target_os = "freebsd",
+))]
+bitflags::bitflags! {
+    /// The flags for a user event.
+    pub struct UserFlags : u32 {
+        /// Ignore the user input flags.
+        const NOINPUT = c::NOTE_FFNOP;
+
+        /// Bitwise AND fflags.
+        const AND = c::NOTE_FFAND;
+
+        /// Bitwise OR fflags.
+        const OR = c::NOTE_FFOR;
+
+        /// Copy fflags.
+        const COPY = c::NOTE_FFCOPY;
+
+        /// Control mask for operations.
+        const CTRLMASK = c::NOTE_FFCTRLMASK;
+
+        /// User defined flags for masks.
+        const UDFMASK = c::NOTE_FFLAGSMASK;
+
+        /// Trigger the event.
+        const TRIGGER = c::NOTE_TRIGGER;
+    }
+}
+
+/// User-defined flags.
+///
+/// Only the lower 24 bits are used in this struct.
+#[repr(transparent)]
+#[cfg(any(
+    target_os = "macos",
+    target_os = "ios",
+    target_os = "watchos",
+    target_os = "dragonfly",
+    target_os = "freebsd",
+))]
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct UserDefinedFlags(u32);
+
+#[cfg(any(
+    target_os = "macos",
+    target_os = "ios",
+    target_os = "watchos",
+    target_os = "dragonfly",
+    target_os = "freebsd",
+))]
+impl UserDefinedFlags {
+    /// Create a new `UserDefinedFlags` from a `u32`.
+    pub fn new(flags: u32) -> Self {
+        Self(flags & EVFILT_USER_FLAGS)
+    }
+
+    /// Get the underlying `u32`.
+    pub fn get(self) -> u32 {
+        self.0
+    }
+}
+
+/// `kqueue()`- Create a new `kqueue` file descriptor.
+pub fn kqueue() -> io::Result<OwnedFd> {
+    syscalls::kqueue()
+}
+
+/// `kevent()`- Wait for events on a `kqueue`.
+pub fn kevent(
+    kqueue: impl AsFd,
+    changelist: &[InEvent<'_>],
```

Comment:
This changelist contains borrowed fds that get registered with the kqueue. What happens if they are closed while they're still registered? And, is there any way to read the fd values back from the kqueue?

Reply:
> What happens if they are closed while they're still registered?

The events will be removed, similarly to how `epoll` does it. From the [man page](https://man.openbsd.org/kqueue.2):

> Calling [close(2)](https://man.openbsd.org/close.2) on a file descriptor will remove any kevents that reference the descriptor.

> And, is there any way to read the fd values back from the kqueue?

In the `FilterKind` structure (accessible via the `OutEvent` structure), the `RawFd` is accessible. I'm not sure if this was the best decision, and I am willing to remove this method from `OutEvent` if you think that it's too much of a footgun.

Reply:
Does it remove the fd from the kqueue on the close of the file descriptor or the file description? The man page says:

> Calling close() on a file descriptor will remove any kevents that reference the descriptor.

makes it sound like it's the file descriptor, but later it says:

> Events which are attached to file descriptors are automatically deleted on the last close of the descriptor.

and the phrase "last close" doesn't make sense for file descriptors because they can only be closed once, but it could make sense for file descriptions.

I ask because on Linux epoll we're using a kind of "weak dup" concept to describe what it does, which works because Linux explicitly works in terms of the file descriptions, making a "weak dup" make sense. But if `kqueue` here is working in terms of the file descriptor, we may need to say something different.

> In the FilterKind structure (accessible via the OutEvent structure), the RawFd is accessible. I'm not sure if this was the best decision, and I am willing to remove this method from OutEvent if you think that it's too much of a footgun.

If the underlying `kqueue` API provides the fd there, then I think Rustix should too. I wasn't happy with the old epoll API being too restrictive, and would like to move toward having APIs providing the full expressivity of the underlying system, even if that requires being `unsafe` in some places.

If `kevent` is causing the queue to hold on to the file *descriptor*, then I don't think we can call it a "weak dup", and my concern is that it's holding onto the file descriptor past the lifetime of the `BorrowedFd`. My initial reaction here is that `EventFilter` should switch from `BorrowedFd` to `RawFd` to reflect the dynamic lifetime of these fds. That would require `kevent` to be `unsafe`; is that too burdensome?

Reply:
> My initial reaction here is that `EventFilter` should switch from `BorrowedFd` to `RawFd` to reflect the dynamic lifetime of these fds. That would require `kevent` to be `unsafe`; is that too burdensome?

I'd be fine with this; my motivating use case (`polling`) is currently stuck on `RawFd` anyways. I agree with being rigorous about borrow checking. Although, in terms of future proofing, if `kqueue` is holding on to the file descriptor, do we want to add a lifetime parameter somewhere in this process? If we add a function that takes `RawFd` we may want to add one that takes `BorrowedFd` in the future, even in the latter is restrictive on lifetimes.

Reply:
From a look at the [`kqueue` source code](https://github.com/freebsd/freebsd-src/blob/dcfa3ee44da2b139f51a8aedb0f55735c6dfe3f3/sys/kern/kern_event.c#L1595-L1598), it looks like it holds the file descriptor itself.

Reply:
Great, thanks for looking into that!